contracts/StateReceiver.sol

```diff
@@ -1,43 +1,106 @@
 pragma solidity ^0.5.11;
 
-import { RLPReader } from "solidity-rlp/contracts/RLPReader.sol";
-
-import { System } from "./System.sol";
+import {RLPReader} from "solidity-rlp/contracts/RLPReader.sol";
+import {System} from "./System.sol";
+import {IStateReceiver} from "./IStateReceiver.sol";
 
 contract StateReceiver is System {
   using RLPReader for bytes;
   using RLPReader for RLPReader.RLPItem;
 
   uint256 public lastStateId;
 
+  bytes32 public failedStateSyncsRoot;
+  mapping(bytes32 => bool) public nullifier;
+
+  mapping(uint256 => bytes) public failedStateSyncs;
+
+  address public rootSetter;
+  uint256 public leafCount;
+  uint256 public replayCount;
+  uint256 public constant TREE_DEPTH = 16;
+
   event StateCommitted(uint256 indexed stateId, bool success);
+  event StateSyncReplay(uint256 indexed stateId);
+
+  constructor(address _rootSetter) public {
+    rootSetter = _rootSetter;
+  }
 
-  function commitState(uint256 syncTime, bytes calldata recordBytes) external onlySystem returns(bool success) {
+  function commitState(uint256 syncTime, bytes calldata recordBytes) external onlySystem returns (bool success) {
     // parse state data
     RLPReader.RLPItem[] memory dataList = recordBytes.toRlpItem().toList();
     uint256 stateId = dataList[0].toUint();
-    require(
-      lastStateId + 1 == stateId,
-      "StateIds are not sequential"
-    );
+    require(lastStateId + 1 == stateId, "StateIds are not sequential");
     lastStateId++;
-
     address receiver = dataList[1].toAddress();
     bytes memory stateData = dataList[2].toBytes();
     // notify state receiver contract, in a non-revert manner
     if (isContract(receiver)) {
       uint256 txGas = 5000000;
+
       bytes memory data = abi.encodeWithSignature("onStateReceive(uint256,bytes)", stateId, stateData);
       // solium-disable-next-line security/no-inline-assembly
       assembly {
         success := call(txGas, receiver, 0, add(data, 0x20), mload(data), 0, 0)
       }
       emit StateCommitted(stateId, success);
+      if (!success) failedStateSyncs[stateId] = abi.encode(receiver, stateData);
     }
   }
 
+  function replayFailedStateSync(uint256 stateId) external {
+    bytes memory stateSyncData = failedStateSyncs[stateId];
+    require(stateSyncData.length != 0, "!found");
+    delete failedStateSyncs[stateId];
+
+    (address receiver, bytes memory stateData) = abi.decode(stateSyncData, (address, bytes));
+    emit StateSyncReplay(stateId);
+    IStateReceiver(receiver).onStateReceive(stateId, stateData); // revertable
+  }
+
+  function setRootAndLeafCount(bytes32 _root, uint256 _leafCount) external {
+    require(msg.sender == rootSetter, "!rootSetter");
+    require(failedStateSyncsRoot == bytes32(0), "!zero");
```

Comment:
`failedStateSyncsRoot`, we can only be configured once & maximum number of nodes  in this tree size are 2 ** TREE_DEPTH -1 (which is `65535`), so we will be able to support only `65535` failed state. 

Did we have any reason for choosing this limit? And is it enough?

Reply:
it's chosen as a reasonable number for dos prevention wrt L1 gas costs, we currently have 31 total failed state syncs (as of last week). happy to increase the depth to 32 w slight gas loss for sanity